src/core/lv_refr.c

```diff
@@ -901,17 +884,46 @@ static void draw_buf_flush(lv_disp_t * disp)
      * and driver is ready to receive the new buffer.
      * If we need to wait here it means that the content of one buffer is being sent to display
      * and other buffer already contains the new rendered image. */
-    if(lv_disp_is_double_buffered(disp)) {
-        while(disp->flushing);
-    }
-
-    disp->flushing = 1;
 
     if(disp->last_area && disp->last_part) disp->flushing_last = 1;
     else disp->flushing_last = 0;
 
     bool flushing_last = disp->flushing_last;
 
+    // This code was moved from _lv_disp_refr_timer as it seem wrong to have a
+    // stall of the program directly after the fluck callback has been called.
+
+    if(lv_disp_is_double_buffered(disp)) {
+        while(disp->flushing);
+        /*If there are 2 buffers swap them. With direct mode swap only on the last area*/
+
+        // This might be wrong and some adjustments might need to be made. I
+        // understaand the buffer is being coppied and I don't think it matters if
+        // it happens before the flush function is called or after because after
+        // all it is a copy being made so long as all rendering to the buffer has
+        // been completed before the copy gets made.
+        if (disp_refr->render_mode == LV_DISP_RENDER_MODE_DIRECT) {
```

Comment:
In direct mode, we needed copy the previously rendered areas to the new buffer to synchronize the content. There a few things to consider:
- We need to copy the invalidated areas of the previous frame, so we can't wait much as the invalidated areas will be clear at the end of `disp_refr_timer`.
- We need to wait until the new frame is flushed, else we will overwrite the being displayed image with the copying. 

It means we need to copy before exiting `disp_refr_timer` and flushing needs to be ready before that. 

We can also save the invalidated areas. This way we can copy at latest before starting the rendering. So before rendering a new frame the other buffer needs to flushed and after that we need to copy previous invalidated areas. Saving the areas will require only a few hundred bytes, so probably we can do this to gain more performance. 

Note that all these copying is required only in direct mode. in full refresh mode we can wait for flushing right before sending a new flush.

Reply:
ok so say you have buffer 1 and buffer 2. there would be no difference if buffer 1 gets copied just before it gets flushed or just after it gets flushed because the only thing that is happening if the buffer data is getting written to the display so the data would not be changing. The only big difference is if it is done after the flush and DMA Memory is being used the flush function is going to exit before the data has been written This is why there was the while loop after the flush exited and before the buffer data was copied. By copying the data before the flush function gets called or expands the amount of time that is being given for the DMA transfer to be done before copying the buffer data again. That time is now being used to render new information to the already copied buffer..  There is still the while loop there in the event it takes longer for the DMA of the data then it does for LVGL to render. But once that buffer has been written to the display the data gets copied to the buffer that just finished being drawn the buffer gets swapped so the buffer that now have the newly copied data is set to the active buffer.

Use this as a rough idea of what is happening.
![image](https://github.com/lvgl/lvgl/assets/10932297/6908cb47-eebc-4229-89ec-440554a83846)

This is what I changed it to.

![image](https://github.com/lvgl/lvgl/assets/10932297/b5fc1a97-8ce5-49e4-955c-80d1610e9fcd)


You can see by placing the buffer copy before the flush function being called will reduce the amount of time the program has to stall or it might remove the need to stall the program entirely depending on the size of the buffer and how the data is being sent. Where as the way it is set up now 100% of the time the program is going to have to stop and wait.

Reply:
>  By copying the data before the flush function gets called or expands the amount of time that is being given for the DMA transfer to be done before copying the buffer data again.

The problem is that you can't do this. Let's say `buf1` is being displayed and we have the next frame to be flushed in `buf2`. The MCU's LCD periphery continuously reads `buf1`. So if we modify it, it will be visible on the screen too.  

That's why we can't copy anything to `buf1` before flushing `buf2`. The best we can do is
1. start flushing `buf2`
2. save the coordinates of areas to copy into `buf1`
3. start copying when the flush is ready but at latest before rendering.

Reply:
> The MCU's LCD periphery continuously reads buf1. So if we modify it, it will be visible on the screen too.

why would it be continuously be reading buf1? isn't that the point of having double buffering? so while one buffer is being written to the display the other can be rendered to?? If that is the case then double buffering when using direct mode would cause a slow down due to the stall and the need to copy the contents of the second buffer onto the first buffer.

You are going to have to elaborate on this a little bit more, I am not understanding what you are saying.

This is the part I am not understanding

> The MCU's LCD periphery continuously reads buf1. So if we modify it, it will be visible on the screen too.

when using double buffering buf1 wouldn't be getting altered while it was being emptied, buf2 would be and then it stalls until the flush function has finished writing buf1. then the data from buf2 gets copied to buf1 and the flush function then gets called to write the data from buf2  and LVGL goes off and renders new data to buf1 while buf2 is being written. Then it circles back around and gets to the stall, this time the program waits until buf2 is written and then the buf1 data gets copied to buf2 and the flush function gets called to write the data in buf1 and LVGL goes off and fills buf2

Reply:
OK and I understand that. which is why there is still the while loop that is in place. so no rendering takes place on the same buffer as what is being displayed and also so the buffer doesn't get changed out while one of them is still being transferred. The exact same mechanics that were there originally is still there it has simply been moved to a different location that allows LVGL to update the second buffer while the first is being emptied where as the way it is set up now is as soon as the flush function exits there is a stall which waits for the buffer to get written. If DMA memory is used the flush function will exit before the data has been written to the display. Only when the buffer has been emptied is when that while loop will exit. This defeats the purpose of using double buffering because the second buffer is not able to be updated while the first is being written to the display.

I can tell you that I do not experience any tearing or anything like that. I have run this test using wxPython as the framework that handles the rendering to the display I am able to simulate what DMA memory transfer would be using what is called a  paint event that Microsoft Windows does when a window needs to be redrawn. This paint event gets created due to the flush function for LVGL telling the window to refresh. once the paint event has drawn the frame buffer data the function gets called letting LVGL know the buffer data has been emptied.